Lib/http/server.py

```diff
@@ -881,6 +878,17 @@ def guess_type(self, path):
         return 'application/octet-stream'
 
 
+def _add_trailing_slash(path):
```

Comment:
I don't like this function. It duplicates what urllib.parse.urlsplit does and is putting logic around parsing URIs in a distant place far from that. It also omits other important behaviors or urlsplit such as control character removal. These are the kinds of things that should not be duplicated logic.

If this is going to exist, it should live in urllib/parse.py.  It should also indicate that it has a dual return value in its name. Returning None for a function claiming to modify something is unusual.  `_add_trailing_slash_or_none` would make it more obvious to someone reading code calling this that a falsey result is expected when no work was found to be done.

like urlsplit, this also deserves a `@functools.lru_cache(typed=True)` if it's going to exist.

Reply:
You are right that it probably should live in `urllib.parse` somewhere.  I renamed it and improved the comment to make it a bit better.  Making it a function exported from `urllib` will take more thought yet.

In RFC 2616, it specifies that the request line has the syntax:
```
Request-URI    = "*" | absoluteURI | abs_path | authority
```
I think `http.server` is only handling the `abs_path` case.  Is it possible to pass absolute URIs to the server?

Reply:
yeah i settled on only handling the abs path case in mine as well - effectively back to the PR the original reporter/author wrote as far as the code fix goes.  when it's got a scheme and netloc that passes through to Location just fine.